src/libraries/System.Text.Json/Common/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,114 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                return name;
+            }
+
+            if (name.Length == 1)
+            {
+                return name.ToLowerInvariant();
+            }
+
+            var result = new ValueStringBuilder(2 * name.Length);
```

Comment:
I'm not sure 2 * length is a reasonable heuristic here, but I wasn't sure how much bigger the "average" property name would be, so I felt just doubling was safer.

Reply:
ABC would turn into a_b_c sot the absolute maximum would be (2 * name.Length) - 1

Reply:
`ABC` should turn into `abc` since the entire string is all the same case; it should be the same case as `ID`

Reply:
so AbCdEf would be ab_cd_ef which gives us name.Length + (name.Length/2) - 1